src/flb_error_reporter.c

```diff
@@ -0,0 +1,225 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <monkey/mk_core/mk_list.h>
+
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_log.h>
+#include <fluent-bit/flb_error_reporter.h>
+
+/* helper function to get int type environment variable*/
+static int getenv_int(const char *name) {
+    char *value, *end;
+    long result;
+
+    value = getenv(name);
+    if (!value)
+        return 0;
+
+    result = strtol(value, &end, 10);
+    if (*end != '\0')
+        return 0;
+
+    return (int)result;
+}
+
+/* create an error reporter*/
+struct flb_error_reporter *flb_error_reporter_create()
+{
+    char *path_var = NULL;
+    int ttl_var, status_message_length;
+    struct flb_error_reporter *error_reporter;
+    FILE *f;
+
+    error_reporter = flb_calloc(1, sizeof(struct flb_error_reporter));
+    if (!error_reporter) {
+        return NULL;
+    }
+
+    /* setup error report file path */
+    path_var = getenv(STATUS_MESSAGE_FILE_PATH_ENV);
+    if (path_var == NULL) {
+        return NULL;
+    }
+    error_reporter->file_path = flb_sds_create(path_var);
+
+    /* setup error reporter message TTL */
+    ttl_var = getenv_int(STATUS_MESSAGE_TTL_ENV);
+    if (ttl_var <= 0) {
+        ttl_var = STATUS_MESSAGE_TTL_DEFAULT;
+    }
+    error_reporter->ttl = ttl_var;
+
+    /* setup error reporter file size */
+    status_message_length = getenv_int(STATUS_MESSAGE_MAX_BYTE_LENGTH_ENV);
+    if(status_message_length <= 0) {
+        status_message_length = STATUS_MESSAGE_MAX_BYTE_LENGTH_DEFAULT;
+    }
+    error_reporter->max_size = status_message_length;
+
+    /* create the message Linked Lists */
+    mk_list_init(&error_reporter->messages);
+    f = fopen(error_reporter->file_path, "w");
+    fclose(f);
+
+    return error_reporter;
+}
+
+/* error reporter write the error message into reporting file and memory*/
+int flb_error_reporter_write(struct flb_error_reporter *error_reporter, char *msg)
+{
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct flb_error_message *message;
+    struct flb_error_message *tmp_message;
+    flb_sds_t buf;
+    int deleted_message_count = 0;
+    FILE *f;
+
+    buf = flb_sds_create(msg);
+
+    /* check if the message is the same with latest one in queue*/
+    if (mk_list_is_empty(&error_reporter->messages) != 0) {
+        tmp_message = mk_list_entry_last(&error_reporter->messages,
+                                         struct flb_error_message, _head);
+        if (tmp_message->len == strlen(buf) &&
+                flb_sds_cmp(tmp_message->data, buf, tmp_message->len) == 0) {
+
+            tmp_message->timestamp = time(NULL);
+            flb_sds_destroy(buf);
+            return 0;
```

Comment:
May be we should chat about this somewhere privately (off github), but why are you only checking the last message? I thought in your design we had said that we'd check for duplicates with any existing messages.

Reply:
As discussed offline, handle only latest message not duplicate is much simple and can works on many situation. This is a trade off. We can go with it and if it's not sufficient we will need to upgrade this part to more complicated check.